File: mcp-code-intelligence-nodejs/src/parsers/languages/typescript-parser.ts
/**
 * KSA-146: TypeScript/JavaScript Language Parser (Modular Architecture).
 * Delegates extraction to specialized sub-extractors for maintainability.
 * Supports both TypeScript (.ts/.tsx) and JavaScript (.js/.jsx/.mjs/.cjs).
 */

import type { ILanguageParser, ParseResult, ExtractedSymbol, ExtractedRelationship, ParseError, SyntaxNode } from '../types.js';
import { findNodes } from '../ast-utils.js';
import { TSSymbolExtractor } from './ts-symbol-extractor.js';
import { TSCallExtractor } from './ts-call-extractor.js';
import { TSImportExtractor } from './ts-import-extractor.js';
import { TSInheritanceExtractor } from './ts-inheritance-extractor.js';

export default class TypeScriptParser implements ILanguageParser {
  readonly languageId: string;
  private parser: any; // web-tree-sitter Parser instance
  private symbolExtractor: TSSymbolExtractor;
  private callExtractor: TSCallExtractor;
  private importExtractor: TSImportExtractor;
  private inheritanceExtractor: TSInheritanceExtractor;

  constructor(parser: any, languageId: string) {
    this.parser = parser;
    this.languageId = languageId;
    this.symbolExtractor = new TSSymbolExtractor();
    this.callExtractor = new TSCallExtractor();
    this.importExtractor = new TSImportExtractor();
    this.inheritanceExtractor = new TSInheritanceExtractor();
  }

  getSupportedExtensions(): string[] {
    return this.languageId === 'typescript'
      ? ['.ts', '.tsx']
      : ['.js', '.jsx', '.mjs', '.cjs'];
  }

  parse(source: string, filePath: string): ParseResult {
    const tree = this.parser.parse(source);
    const rootNode = tree.rootNode;

    // Collect parse errors
    const errors: ParseError[] = [];
    if (rootNode.hasError()) {
      const errorNodes = findNodes(rootNode, 'ERROR');
      for (const node of errorNodes.slice(0, 10)) {
        errors.push({
          message: 'Parse error',
          line: node.startPosition.row + 1,
          column: node.startPosition.column,
        });
      }
    }

    // Extract symbols (functions, classes, interfaces, etc.)
    const relationships: ExtractedRelationship[] = [];
    const symbols = this.symbolExtractor.extract(rootNode, source, filePath, relationships);

    // Extract call relationships
    const calls = this.callExtractor.extract(rootNode, source, filePath, symbols);

    // Extract import relationships
    const imports = this.importExtractor.extract(rootNode, source, filePath);

    // Extract inheritance relationships
    const inheritance = this.inheritanceExtractor.extract(rootNode, source, filePath, symbols);

    // Merge all relationships
    const allRelationships = [...relationships, ...calls, ...imports, ...inheritance];

    return { symbols, relationships: allRelationships, errors };
  }
}

File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/parsers/languages/TypeScriptParser.kt
/**
 * KSA-172: TypeScript/JavaScript Language Parser.
 * Regex-based extraction matching nodejs tree-sitter output structure.
 */
package com.codeintel.parsers.languages

import com.codeintel.parsers.*

class TypeScriptParser(languageId: String = "typescript") : BaseLanguageParser(languageId) {
    override fun getSupportedExtensions(): List<String> =
        if (languageId == "typescript") listOf(".ts", ".tsx") else listOf(".js", ".jsx", ".mjs", ".cjs")

    override fun parse(source: String, filePath: String): ParseResult {
        val symbols = mutableListOf<ExtractedSymbol>()
        val rels = mutableListOf<ExtractedRelationship>()
        val lines = source.lines()
        val importRe = Regex("""import\s+.*?from\s+['"]([^'"]+)['"]""")
        val funcRe = Regex("""^(export\s+)?(async\s+)?function\s*\*?\s+(\w+)\s*\(([^)]*)\)""")
        val classRe = Regex("""^(export\s+)?(abstract\s+)?class\s+(\w+)""")
        val ifaceRe = Regex("""^(export\s+)?interface\s+(\w+)""")
        val enumRe = Regex("""^(export\s+)?enum\s+(\w+)""")
        for ((i, line) in lines.withIndex()) {
            val t = line.trim(); val ln = i + 1
            importRe.find(t)?.let { rels.add(buildRelationship("__file__", it.groupValues[1], "imports", filePath, ln)) }
            funcRe.find(t)?.let { m ->
                symbols.add(buildSymbol(m.groupValues[3], "function", filePath, ln, ln, "${if (m.groupValues[2].isNotBlank()) "async " else ""}function ${m.groupValues[3]}(${m.groupValues[4]})", parameters="(${m.groupValues[4]})", isAsync=m.groupValues[2].isNotBlank(), isExported=m.groupValues[1].isNotBlank()))
            }
            classRe.find(t)?.let { m ->
                symbols.add(buildSymbol(m.groupValues[3], "class", filePath, ln, ln, "${if (m.groupValues[2].isNotBlank()) "abstract " else ""}class ${m.groupValues[3]}", modifiers=if (m.groupValues[2].isNotBlank()) listOf("abstract") else null, isExported=m.groupValues[1].isNotBlank()))
            }
            ifaceRe.find(t)?.let { m -> symbols.add(buildSymbol(m.groupValues[2], "interface", filePath, ln, ln, "interface ${m.groupValues[2]}", isExported=m.groupValues[1].isNotBlank())) }
            enumRe.find(t)?.let { m -> symbols.add(buildSymbol(m.groupValues[2], "enum", filePath, ln, ln, "enum ${m.groupValues[2]}", isExported=m.groupValues[1].isNotBlank())) }
        }
        return ParseResult(symbols, rels)
    }
}
